src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/NuGetProjectService.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.PackageManagement.VisualStudio;
+using NuGet.Packaging;
+using NuGet.ProjectManagement;
+using NuGet.ProjectManagement.Projects;
+using NuGet.VisualStudio.Contracts;
+
+namespace NuGet.VisualStudio.Implementation.Extensibility
+{
+    internal class NuGetProjectService : INuGetProjectService
+    {
+        private readonly IVsSolutionManager _solutionManager;
+
+        public NuGetProjectService(IVsSolutionManager solutionManager)
+        {
+            _solutionManager = solutionManager ?? throw new ArgumentNullException(nameof(solutionManager));
+        }
+
+        public async Task<InstalledPackagesResult> GetInstalledPackagesAsync(Guid projectId, CancellationToken cancellationToken)
+        {
+            // Just in case we're on the UI thread, switch to background thread. Very low cost (does not schedule new task) if already on background thread.
+            await TaskScheduler.Default;
+
+            NuGetProject project = await _solutionManager.GetNuGetProjectAsync(projectId.ToString());
+            if (project == null)
+            {
+                return NuGetContractsFactory.CreateInstalledPackagesResult(InstalledPackageResultStatus.ProjectNotReady, packages: null);
+            }
+
+            var status = InstalledPackageResultStatus.Successful;
+            List<NuGetInstalledPackage> installedPackages;
+
+            if (project is BuildIntegratedNuGetProject packageReferenceProject)
+            {
+                var cacheContext = new DependencyGraphCacheContext();
+                var (packageSpecs, messages) = await packageReferenceProject.GetPackageSpecsAndAdditionalMessagesAsync(cacheContext);
+                if (messages?.Any(m => m.Level == LogLevel.Error) == true)
+                {
+                    status = InstalledPackageResultStatus.ProjectInvalid;
+                }
```

Comment:
Maybe have an `else` here to avoid looking at packageSpecs when we know there's an Error? Seems if the Project is Invalid, we don't need to do any more work.

Reply:
We might still have some package information about projects with nomination errors. I think it would be good to return what we can, while still telling the caller that this project is not in a good state.